### PR TITLE
feat: automatically resolve cluster in `talosctl` calls

### PR DESCRIPTION
**Pull request by Unix4ever**
`talosctl --cluster` flag is now optional, Omni will automatically resolve the cluster if the machine is a part of one.

Fixes: https://github.com/siderolabs/omni/issues/620

### CODE REVIEW

File: internal/backend/grpc/router/router.go [hunk 1]
```diff
@@ -173,10 +173,28 @@ func (r *Router) Director(ctx context.Context, fullMethodName string) (proxy.Mod
 func (r *Router) getTalosBackend(ctx context.Context, md metadata.MD) ([]proxy.Backend, error) {
 	clusterName := getClusterName(md)
 
+	nodeID := getNodeID(md)
+
 	id := fmt.Sprintf("cluster-%s", clusterName)
 
 	if clusterName == "" {
-		id = fmt.Sprintf("machine-%s", getNodeID(md))
+		machineSetNode, err := safe.ReaderGetByID[*omni.MachineSetNode](actor.MarkContextAsInternalActor(ctx), r.cosiState, nodeID)
```
Comment (smira) [inline internal/backend/grpc/router/router.go:181]: should we add an integration test for all the ways we have right now?

File: internal/backend/grpc/router/router.go [hunk 2]
```diff
@@ -173,10 +173,28 @@ func (r *Router) Director(ctx context.Context, fullMethodName string) (proxy.Mod
 func (r *Router) getTalosBackend(ctx context.Context, md metadata.MD) ([]proxy.Backend, error) {
 	clusterName := getClusterName(md)
 
+	nodeID := getNodeID(md)
```
Comment (smira) [inline internal/backend/grpc/router/router.go:176]: it's not clear, does it work only with machine UUIDs?

File: internal/backend/grpc/router/router.go (hunk 1 above)
Comment (smira) [reply internal/backend/grpc/router/router.go:181]: we already had one, it just might need to be updated

File: internal/backend/grpc/router/router.go [hunk 3]
```diff
@@ -173,10 +173,28 @@ func (r *Router) Director(ctx context.Context, fullMethodName string) (proxy.Mod
...
+			var ok bool
+
+			clusterName, ok = machineSetNode.Metadata().Labels().Get(omni.LabelCluster)
+			if !ok {
+				return nil, fmt.Errorf("failed to resolve machine cluster relation: no cluster name in the machine set node %q", id)
+			}
+
+			id = fmt.Sprintf("cluster-%s", clusterName)
+		} else {
+			id = fmt.Sprintf("machine-%s", nodeID)
```
Comment (DmitriyMV) [inline internal/backend/grpc/router/router.go:196]:
```suggestion
			id ="machine-" + nodeID
```

Minor nit: perhaps this is clearer?

File: internal/backend/grpc/router/router.go (hunk 2 above)
Comment (Unix4ever) [reply internal/backend/grpc/router/router.go:176]: that's a great question. I'm afraid that with the hostname the resolution might be ambiguous.
I think I can make it work: we can show the error if the resolution gives several clusters.

File: internal/backend/grpc/router/router.go (hunk 1 above)
Comment (Unix4ever) [reply internal/backend/grpc/router/router.go:181]: yeah, will do that

File: internal/backend/dns/service.go [hunk 4]
```diff
@@ -55,13 +58,41 @@ func (i Info) GetAddress() string {
...
+func (m resolverMap) remove(key string, id resource.ID) {
+	ids, ok := m[key]
+	if !ok {
+		return
+	}
+
+	ids = slices.DeleteFunc(ids, func(item string) bool { return item == id })
+	if len(ids) == 0 {
+		delete(m, key)
+	}
```
Comment (DmitriyMV) [inline internal/backend/dns/service.go:84]:
```suggestion
	if len(ids) == 0 {
		delete(m, key)
	} else {
		m[key] = ids	
	}
```

**Review by DmitriyMV — APPROVED** [review]

File: internal/backend/dns/service.go (hunk 4 above)
Comment (Unix4ever) [reply internal/backend/dns/service.go:84]: good catch!

**Pull request merged**